File: backend/app/services/trade_projection.py
"""
Trade Material Projection - Core Data Models

This module defines the data structures for trade material projections.
Key principle: Clear separation between ground truth (measured) and projections (estimates).

Ground Truth (AufmassItem):
- Derived directly from extraction results
- 100% traceable to source measurements
- Formulas are deterministic

Projections (ProjectedMaterial):
- Estimates based on rules or LLM suggestions
- Always labeled as estimates
- Include assumptions and confidence scores
"""

from dataclasses import dataclass, field
from typing import List, Dict, Optional, Any, Tuple
from enum import Enum
from uuid import uuid4
from datetime import datetime


class TradeType(str, Enum):
    """Available construction trades for material projection."""
    SCAFFOLDING = "scaffolding"      # Gerüstbau
    DRYWALL = "drywall"              # Trockenbau
    SCREED = "screed"                # Estrich
    FLOOR_FINISH = "floor_finish"    # Oberbelag
    WATERPROOFING = "waterproofing"  # Abdichtung


class ProjectionMethod(str, Enum):
    """Method used to generate projection."""
    RULE_BASED = "rule_based"
    LLM_ASSISTED = "llm_assisted"


class ConfidenceLevel(str, Enum):
    """Confidence level categories."""
    HIGH = "high"      # 0.8-1.0 - Direct measurement-based
    MEDIUM = "medium"  # 0.5-0.8 - Standard formulas with assumptions
    LOW = "low"        # 0.0-0.5 - Estimates or LLM suggestions


# Trade metadata for UI and validation
TRADE_METADATA = {
    TradeType.SCAFFOLDING: {
        "name_de": "Gerüstbau",
        "name_en": "Scaffolding",
        "icon": "scaffold",
        "required_params": ["scaffold_height_m"],
        "optional_params": ["scaffold_type", "waste_factor"],
        "uses_perimeter": True,
        "uses_area": False,
    },
    TradeType.DRYWALL: {
        "name_de": "Trockenbau",
        "name_en": "Drywall",
        "icon": "wall",
        "required_params": ["wall_height_m"],
        "optional_params": ["drywall_system", "stud_spacing_mm", "waste_factor"],
        "uses_perimeter": True,
        "uses_area": False,
    },
    TradeType.SCREED: {
        "name_de": "Estrich",
        "name_en": "Screed",
        "icon": "layers",
        "required_params": ["screed_thickness_mm"],
        "optional_params": ["screed_type", "waste_factor"],
        "uses_perimeter": True,  # For edge strips
        "uses_area": True,
    },
    TradeType.FLOOR_FINISH: {
        "name_de": "Oberbelag",
        "name_en": "Floor Finish",
        "icon": "grid",
        "required_params": ["finish_type"],
        "optional_params": ["waste_factor"],
        "uses_perimeter": True,  # For skirting
        "uses_area": True,
    },
    TradeType.WATERPROOFING: {
        "name_de": "Abdichtung",
        "name_en": "Waterproofing",
        "icon": "droplet",
        "required_params": [],
        "optional_params": ["waterproofing_type", "wall_height_m", "waste_factor"],
        "uses_perimeter": True,
        "uses_area": True,
    },
}


def _generate_projection_id() -> str:
    """Generate a unique projection ID."""
    return f"proj_{uuid4().hex[:12]}"


def _generate_item_id(prefix: str = "item") -> str:
    """Generate a unique item ID with prefix."""
    return f"{prefix}_{uuid4().hex[:8]}"


def _get_confidence_level(confidence: float) -> ConfidenceLevel:
    """Map confidence score to level."""
    if confidence >= 0.8:
        return ConfidenceLevel.HIGH
    elif confidence >= 0.5:
        return ConfidenceLevel.MEDIUM
    return ConfidenceLevel.LOW


@dataclass
class GroundTruthMeasurement:
    """
    Reference to an extracted measurement (immutable ground truth).

    This links projections back to the original extraction results,
    ensuring full traceability.
    """
    measurement_id: str
    source_type: str          # "extraction_result" | "gewerk_result"
    source_field: str         # e.g., "total_area_m2", "rooms[].perimeter_m"
    value: float
    unit: str                 # "m2", "m", "count"
    source_page: Optional[int] = None
    source_room_id: Optional[str] = None
    source_text: Optional[str] = None  # Original text from PDF

    def to_dict(self) -> Dict[str, Any]:
        return {
            "measurement_id": self.measurement_id,
            "source_type": self.source_type,
            "source_field": self.source_field,
            "value": round(self.value, 4),
            "unit": self.unit,
            "source_page": self.source_page,
            "source_room_id": self.source_room_id,
            "source_text": self.source_text,
        }


@dataclass
class AufmassItem:
    """
    Calculated item based on measured quantities (ground truth formulas).

    These are deterministic calculations derived from extraction results.
    Every item traces back to source measurements.
    """
    item_id: str
    position: str             # LV position number (e.g., "01.01")
    description: str
    quantity: float
    unit: str

    # Traceability - required
    derived_from: List[GroundTruthMeasurement] = field(default_factory=list)
    formula: str = ""         # e.g., "total_area_m2 * 1.10"
    formula_description: str = ""  # Human readable explanation

    # Optional pricing
    unit_price_eur: Optional[float] = None
    total_price_eur: Optional[float] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "item_id": self.item_id,
            "position": self.position,
            "description": self.description,
            "quantity": round(self.quantity, 2),
            "unit": self.unit,
            "derived_from": [m.to_dict() for m in self.derived_from],
            "formula": self.formula,
            "formula_description": self.formula_description,
            "unit_price_eur": self.unit_price_eur,
            "total_price_eur": round(self.total_price_eur, 2) if self.total_price_eur else None,
            "is_ground_truth": True,  # Always true for AufmassItem
        }


@dataclass
class ProjectedMaterial:
    """
    Estimated material requirement (clearly labeled as projection).

    These are NOT ground truth - they are estimates based on rules or LLM.
    Always includes assumptions and confidence scores.
    """
    material_id: str
    name: str                 # e.g., "Estrich FE 50", "CW 75 Profile"
    quantity: float
    unit: str

    # Projection metadata - required
    method: ProjectionMethod
    confidence: float         # 0.0 - 1.0
    assumptions: List[str] = field(default_factory=list)

    # Source traceability
    derived_from_aufmass: Optional[str] = None  # Links to AufmassItem.item_id

    # Optional: LLM suggestion details (only if method == LLM_ASSISTED)
    llm_suggestion: Optional[Dict[str, Any]] = None

    # User can override the quantity
    user_override: Optional[float] = None
    user_override_reason: Optional[str] = None

    @property
    def effective_quantity(self) -> float:
        """Return user override if set, otherwise calculated quantity."""
        return self.user_override if self.user_override is not None else self.quantity

    @property
    def confidence_level(self) -> ConfidenceLevel:
        """Get confidence level category."""
        return _get_confidence_level(self.confidence)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "material_id": self.material_id,
            "name": self.name,
            "quantity": round(self.quantity, 2),
            "effective_quantity": round(self.effective_quantity, 2),
            "unit": self.unit,
            "method": self.method.value,
            "confidence": round(self.confidence, 2),
            "confidence_level": self.confidence_level.value,
            "assumptions": self.assumptions,
            "derived_from_aufmass": self.derived_from_aufmass,
            "llm_suggestion": self.llm_suggestion,
            "user_override": self.user_override,
            "user_override_reason": self.user_override_reason,
            "is_estimate": True,  # Always true for ProjectedMaterial
        }


@dataclass
class TradeParams:
    """
    User-provided parameters for trade calculations.

    Different trades require different parameters.
    """
    trade_type: TradeType

    # Common params
    waste_factor: float = 1.10      # 10% default waste

    # Scaffolding specific
    scaffold_height_m: Optional[float] = None
    scaffold_type: str = "standard"  # "standard", "rollgeruest", "fassade"

    # Drywall specific
    wall_height_m: Optional[float] = None
    drywall_system: str = "single"   # "single", "double", "fire_rated"
    stud_spacing_mm: int = 625       # Standard spacing

    # Screed specific
    screed_type: str = "ct"          # "ct" (cement), "ca" (anhydrite), "ma" (mastic)
    screed_thickness_mm: int = 50

    # Floor finish specific
    finish_type: str = "laminate"    # "laminate", "parquet", "tile", "carpet"

    # Waterproofing specific
    waterproofing_type: str = "liquid"  # "liquid", "membrane", "bitumen"
    waterproofing_wall_height_m: float = 2.0  # Height to apply in wet areas

    def validate(self) -> List[str]:
        """Validate parameters for the trade type. Returns list of errors."""
        errors = []
        metadata = TRADE_METADATA.get(self.trade_type)

        if not metadata:
            errors.append(f"Unknown trade type: {self.trade_type}")
            return errors

        for param in metadata["required_params"]:
            value = getattr(self, param, None)
            if value is None:
                errors.append(f"Missing required parameter: {param}")
            elif isinstance(value, (int, float)) and value <= 0:
                errors.append(f"Parameter {param} must be positive")

        if self.waste_factor < 1.0 or self.waste_factor > 2.0:
            errors.append("waste_factor should be between 1.0 and 2.0")

        return errors

    def to_dict(self) -> Dict[str, Any]:
        return {
            "trade_type": self.trade_type.value,
            "waste_factor": self.waste_factor,
            "scaffold_height_m": self.scaffold_height_m,
            "scaffold_type": self.scaffold_type,
            "wall_height_m": self.wall_height_m,
            "drywall_system": self.drywall_system,
            "stud_spacing_mm": self.stud_spacing_mm,
            "screed_type": self.screed_type,
            "screed_thickness_mm": self.screed_thickness_mm,
            "finish_type": self.finish_type,
            "waterproofing_type": self.waterproofing_type,
            "waterproofing_wall_height_m": self.waterproofing_wall_height_m,
        }


@dataclass
class MaterialProjectionResult:
    """
    Complete result for a trade material projection.

    Contains both ground truth (aufmass_items) and estimates (projected_materials).
    """
    projection_id: str
    trade_type: TradeType
    source_extraction_id: str
    processed_at: str
    status: str = "ok"  # "ok", "partial", "error"

    # Parameters used
    params: Optional[TradeParams] = None

    # Results
    aufmass_items: List[AufmassItem] = field(default_factory=list)
    projected_materials: List[ProjectedMaterial] = field(default_factory=list)

    # Totals
    total_aufmass_quantity: float = 0.0
    total_estimated_cost_eur: Optional[float] = None

    # Metadata
    errors: List[str] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)

    # Legal disclaimer - always included
    disclaimer: str = (
        "Projected materials are estimates based on standard formulas and assumptions. "
        "Actual quantities may vary based on site conditions, material specifications, "
        "and construction methods. These projections should be verified by a qualified professional."
    )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "projection_id": self.projection_id,
            "trade_type": self.trade_type.value,
            "trade_name_de": TRADE_METADATA[self.trade_type]["name_de"],
            "source_extraction_id": self.source_extraction_id,
            "processed_at": self.processed_at,
            "status": self.status,
            "params": self.params.to_dict() if self.params else None,
            "aufmass_items": [i.to_dict() for i in self.aufmass_items],
            "projected_materials": [m.to_dict() for m in self.projected_materials],
            "total_aufmass_quantity": round(self.total_aufmass_quantity, 2),
            "total_estimated_cost_eur": round(self.total_estimated_cost_eur, 2) if self.total_estimated_cost_eur else None,
            "errors": self.errors,
            "warnings": self.warnings,
            "disclaimer": self.disclaimer,
            "summary": {
                "aufmass_count": len(self.aufmass_items),
                "materials_count": len(self.projected_materials),
                "has_errors": len(self.errors) > 0,
                "has_warnings": len(self.warnings) > 0,
            },
        }


# Helper functions for creating projection results

def create_ground_truth_measurement(
    source_field: str,
    value: float,
    unit: str,
    source_type: str = "extraction_result",
    source_page: Optional[int] = None,
    source_room_id: Optional[str] = None,
    source_text: Optional[str] = None,
) -> GroundTruthMeasurement:
    """Create a GroundTruthMeasurement with auto-generated ID."""
    return GroundTruthMeasurement(
        measurement_id=_generate_item_id("meas"),
        source_type=source_type,
        source_field=source_field,
        value=value,
        unit=unit,
        source_page=source_page,
        source_room_id=source_room_id,
        source_text=source_text,
    )


def create_aufmass_item(
    position: str,
    description: str,
    quantity: float,
    unit: str,
    derived_from: List[GroundTruthMeasurement],
    formula: str,
    formula_description: str = "",
) -> AufmassItem:
    """Create an AufmassItem with auto-generated ID."""
    return AufmassItem(
        item_id=_generate_item_id("auf"),
        position=position,
        description=description,
        quantity=quantity,
        unit=unit,
        derived_from=derived_from,
        formula=formula,
        formula_description=formula_description,
    )


def create_projected_material(
    name: str,
    quantity: float,
    unit: str,
    confidence: float,
    assumptions: List[str],
    method: ProjectionMethod = ProjectionMethod.RULE_BASED,
    derived_from_aufmass: Optional[str] = None,
) -> ProjectedMaterial:
    """Create a ProjectedMaterial with auto-generated ID."""
    return ProjectedMaterial(
        material_id=_generate_item_id("mat"),
        name=name,
        quantity=quantity,
        unit=unit,
        method=method,
        confidence=confidence,
        assumptions=assumptions,
        derived_from_aufmass=derived_from_aufmass,
    )


def create_projection_result(
    trade_type: TradeType,
    source_extraction_id: str,
    params: TradeParams,
    aufmass_items: List[AufmassItem],
    projected_materials: List[ProjectedMaterial],
    warnings: Optional[List[str]] = None,
    errors: Optional[List[str]] = None,
) -> MaterialProjectionResult:
    """Create a MaterialProjectionResult with auto-generated ID and timestamp."""

    # Calculate total aufmass quantity (sum of primary unit)
    total_qty = sum(item.quantity for item in aufmass_items)

    return MaterialProjectionResult(
        projection_id=_generate_projection_id(),
        trade_type=trade_type,
        source_extraction_id=source_extraction_id,
        processed_at=datetime.utcnow().isoformat() + "Z",
        status="error" if errors else ("partial" if warnings else "ok"),
        params=params,
        aufmass_items=aufmass_items,
        projected_materials=projected_materials,
        total_aufmass_quantity=total_qty,
        warnings=warnings or [],
        errors=errors or [],
    )

File: frontend/lib/trade-projection.ts
/**
 * Trade Material Projection - TypeScript Types and API Client
 *
 * This module provides types and API functions for the trade material
 * projection system. It separates ground truth (measured) from projections (estimates).
 */

// ==================
// TYPES
// ==================

export type TradeType =
  | 'scaffolding'
  | 'drywall'
  | 'screed'
  | 'floor_finish'
  | 'waterproofing';

export type ProjectionMethod = 'rule_based' | 'llm_assisted';

export type ConfidenceLevel = 'high' | 'medium' | 'low';

/**
 * Reference to a measurement from the extraction result.
 * Used for traceability.
 */
export interface GroundTruthMeasurement {
  measurement_id: string;
  source_type: string;
  source_field: string;
  value: number;
  unit: string;
  source_page?: number;
  source_room_id?: string;
  source_text?: string;
}

/**
 * Aufmass item - calculated from measured quantities.
 * Always ground truth with full traceability.
 */
export interface AufmassItem {
  item_id: string;
  position: string;
  description: string;
  quantity: number;
  unit: string;
  derived_from: GroundTruthMeasurement[];
  formula: string;
  formula_description: string;
  unit_price_eur?: number;
  total_price_eur?: number;
  is_ground_truth: true;
}

/**
 * Projected material - estimate with assumptions and confidence.
 * Always clearly labeled as estimate.
 */
export interface ProjectedMaterial {
  material_id: string;
  name: string;
  quantity: number;
  effective_quantity: number;
  unit: string;
  method: ProjectionMethod;
  confidence: number;
  confidence_level: ConfidenceLevel;
  assumptions: string[];
  derived_from_aufmass?: string;
  llm_suggestion?: Record<string, unknown>;
  user_override?: number;
  user_override_reason?: string;
  is_estimate: true;
}

/**
 * Parameters for projection calculation.
 */
export interface ProjectionParams {
  waste_factor?: number;

  // Scaffolding
  scaffold_height_m?: number;
  scaffold_type?: 'standard' | 'rollgeruest' | 'fassade';

  // Drywall
  wall_height_m?: number;
  drywall_system?: 'single' | 'double' | 'fire_rated';
  stud_spacing_mm?: number;

  // Screed
  screed_type?: 'ct' | 'ca' | 'ma';
  screed_thickness_mm?: number;

  // Floor finish
  finish_type?: 'laminate' | 'parquet' | 'tile' | 'carpet';

  // Waterproofing
  waterproofing_type?: 'liquid' | 'membrane' | 'bitumen';
  waterproofing_wall_height_m?: number;
}

/**
 * Request for material projection.
 */
export interface ProjectionRequest {
  extraction_result: Record<string, unknown>;
  trade_type: TradeType;
  params: ProjectionParams;
  use_llm?: boolean;
}

/**
 * Result summary.
 */
export interface ProjectionSummary {
  aufmass_count: number;
  materials_count: number;
  has_errors: boolean;
  has_warnings: boolean;
}

/**
 * Complete projection result.
 */
export interface ProjectionResult {
  projection_id: string;
  trade_type: TradeType;
  trade_name_de: string;
  source_extraction_id: string;
  processed_at: string;
  status: 'ok' | 'partial' | 'error';
  params: ProjectionParams;
  aufmass_items: AufmassItem[];
  projected_materials: ProjectedMaterial[];
  total_aufmass_quantity: number;
  total_estimated_cost_eur?: number;
  errors: string[];
  warnings: string[];
  disclaimer: string;
  summary: ProjectionSummary;
}

/**
 * Trade information from API.
 */
export interface TradeInfo {
  type: TradeType;
  name_de: string;
  name_en: string;
  icon: string;
  required_params: string[];
  optional_params: string[];
  uses_perimeter: boolean;
  uses_area: boolean;
}

// ==================
// API CLIENT
// ==================

const API_BASE = process.env.NEXT_PUBLIC_SNAPGRID_API_URL || 'http://localhost:8000';

/**
 * Get list of available trades.
 */
export async function getAvailableTrades(): Promise<TradeInfo[]> {
  const response = await fetch(`${API_BASE}/api/v1/projections/trades`);

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to fetch trades' }));
    throw new Error(error.detail || 'Failed to fetch available trades');
  }

  const data = await response.json();
  return data.trades;
}

/**
 * Get detailed information about a specific trade.
 */
export async function getTradeInfo(tradeType: TradeType): Promise<TradeInfo> {
  const response = await fetch(`${API_BASE}/api/v1/projections/trades/${tradeType}`);

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to fetch trade info' }));
    throw new Error(error.detail || `Failed to fetch info for trade: ${tradeType}`);
  }

  return response.json();
}

/**
 * Calculate material projection for a trade.
 */
export async function calculateProjection(
  request: ProjectionRequest
): Promise<ProjectionResult> {
  const response = await fetch(`${API_BASE}/api/v1/projections/calculate`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Projection calculation failed' }));
    throw new Error(error.detail || 'Failed to calculate projection');
  }

  return response.json();
}

/**
 * Calculate projections for multiple trades at once.
 */
export async function calculateBatchProjections(
  extractionResult: Record<string, unknown>,
  tradeTypes: TradeType[],
  params?: ProjectionParams
): Promise<Record<TradeType, ProjectionResult | { status: 'error'; error: string }>> {
  const url = new URL(`${API_BASE}/api/v1/projections/calculate/batch`);
  tradeTypes.forEach(t => url.searchParams.append('trade_types', t));

  const response = await fetch(url.toString(), {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      extraction_result: extractionResult,
      params: params || {},
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Batch projection failed' }));
    throw new Error(error.detail || 'Failed to calculate batch projections');
  }

  const data = await response.json();
  return data.results;
}

/**
 * Export projection to JSON format.
 */
export async function exportProjectionJson(
  projection: ProjectionResult
): Promise<Blob> {
  const response = await fetch(`${API_BASE}/api/v1/projections/export/json`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(projection),
  });

  if (!response.ok) {
    throw new Error('Failed to export projection');
  }

  const data = await response.json();
  return new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
}

// ==================
// UTILITIES
// ==================

/**
 * Trade display information for UI.
 */
export const TRADE_DISPLAY = {
  scaffolding: {
    name: 'Gerüstbau',
    description: 'Facade scaffolding and mobile scaffolds',
    icon: '🏗️',
    color: 'bg-orange-500',
  },
  drywall: {
    name: 'Trockenbau',
    description: 'Gypsum boards, profiles, and fasteners',
    icon: '🧱',
    color: 'bg-gray-500',
  },
  screed: {
    name: 'Estrich',
    description: 'Cement and anhydrite screed',
    icon: '🔲',
    color: 'bg-stone-500',
  },
  floor_finish: {
    name: 'Oberbelag',
    description: 'Laminate, parquet, tile, carpet',
    icon: '🏠',
    color: 'bg-amber-500',
  },
  waterproofing: {
    name: 'Abdichtung',
    description: 'Waterproofing for wet rooms',
    icon: '💧',
    color: 'bg-blue-500',
  },
} as const;

/**
 * Get confidence badge color based on level.
 */
export function getConfidenceColor(level: ConfidenceLevel): string {
  switch (level) {
    case 'high':
      return 'bg-green-500/20 text-green-400 border-green-500/30';
    case 'medium':
      return 'bg-yellow-500/20 text-yellow-400 border-yellow-500/30';
    case 'low':
      return 'bg-red-500/20 text-red-400 border-red-500/30';
    default:
      return 'bg-gray-500/20 text-gray-400 border-gray-500/30';
  }
}

/**
 * Get confidence badge label.
 */
export function getConfidenceLabel(level: ConfidenceLevel): string {
  switch (level) {
    case 'high':
      return 'Hoch';
    case 'medium':
      return 'Mittel';
    case 'low':
      return 'Niedrig';
    default:
      return 'Unbekannt';
  }
}

/**
 * Format number with German locale.
 */
export function formatQuantity(value: number, decimals: number = 2): string {
  return value.toLocaleString('de-DE', {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  });
}

/**
 * Download blob as file.
 */
export function downloadBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

// ==================
// DRYWALL SYMBOL DETECTION (Plankopf-based)
// ==================

/**
 * Legend symbol from Plankopf parsing.
 */
export interface LegendSymbol {
  symbol_id: string;
  label: string;
  label_normalized: string;
  material_type: string | null;
  pattern_info: {
    pattern_type: string;
    stroke_color: number[] | null;
    fill_color: number[] | null;
    hatching_angle: number | null;
    hatching_spacing: number | null;
    line_count: number;
  };
  confidence: number;
}

/**
 * Detected drywall segment.
 */
export interface DrywallSegment {
  segment_id: string;
  bbox: { x0: number; y0: number; x1: number; y1: number };
  length_m: number;
  area_m2: number;
  wall_height_m: number;
  confidence: number;
  page_number: number;
}

/**
 * Detection result for a single drywall type.
 */
export interface DrywallResultItem {
  detection_id: string;
  material_label: string;
  material_type: string;
  segments: DrywallSegment[];
  total_count: number;
  total_length_m: number;
  total_area_m2: number;
  wall_height_m: number;
  scale: string;
  confidence: number;
  pages_analyzed: number[];
  assumptions: string[];
  warnings: string[];
}

/**
 * Full drywall detection response from Plankopf-based analysis.
 */
export interface DrywallDetectionResult {
  extraction_id: string;
  filename: string;
  page_count: number;
  status: 'ok' | 'partial' | 'error';

  // Plankopf info
  plankopf_found: boolean;
  plankopf_page: number | null;
  legend_symbols: LegendSymbol[];

  // Detection results
  drywall_results: DrywallResultItem[];

  // Totals
  grand_total_length_m: number;
  grand_total_area_m2: number;
  grand_total_segments: number;

  // Settings
  scale: string;
  wall_height_m: number;

  // Metadata
  processed_at: string;
  errors: string[];
  warnings: string[];
}

/**
 * Detect drywall using Plankopf symbol matching.
 * This reads the legend/title block to learn what patterns represent drywall,
 * then scans the drawing for matching patterns.
 */
export async function detectDrywallFromSymbols(
  file: File,
  options?: {
    wall_height_m?: number;
    target_label?: string;
    scale?: string;
    page_numbers?: number[];
  }
): Promise<DrywallDetectionResult> {
  const formData = new FormData();
  formData.append('file', file);

  const params = new URLSearchParams();
  if (options?.wall_height_m) {
    params.append('wall_height_m', options.wall_height_m.toString());
  }
  if (options?.target_label) {
    params.append('target_label', options.target_label);
  }
  if (options?.scale) {
    params.append('scale', options.scale);
  }
  if (options?.page_numbers && options.page_numbers.length > 0) {
    params.append('page_numbers', options.page_numbers.join(','));
  }

  const url = `${API_BASE}/api/v1/drywall-detection/from-symbols${params.toString() ? '?' + params.toString() : ''}`;
  const response = await fetch(url, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Detection failed' }));
    throw new Error(error.detail || 'Failed to detect drywall from symbols');
  }

  return response.json();
}

/**
 * Analyze only the Plankopf/legend without full detection.
 * Use this to preview what symbols are available.
 */
export async function analyzePlankopfLegend(
  file: File,
  pageNumber: number = 0
): Promise<{
  plankopf_found: boolean;
  page_number: number;
  symbols: LegendSymbol[];
  drywall_symbols: LegendSymbol[];
  warnings: string[];
}> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(
    `${API_BASE}/api/v1/drywall-detection/analyze-legend?page_number=${pageNumber}`,
    {
      method: 'POST',
      body: formData,
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Legend analysis failed' }));
    throw new Error(error.detail || 'Failed to analyze legend');
  }

  return response.json();
}
